File: Graphs/LC_210_Course_Schedule_II.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.18

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        # practice topology sorting
        n = numCourses
        indegree = [0] * n
        mapping = collections.defaultdict(list)
        total_edges, removed_edges = 0, 0
        for start, end  in prerequisites:
            indegree[end] += 1
            mapping[start].append(end)
            total_edges += 1
        
        q = collections.deque()
        for i in range(n):
            if indegree[i] == 0:
                q.append(i)
        ans = []
        while q:
            size = len(q)
            for _ in range(size):
                cur = q.popleft()
                ans.append(cur)
                for nxt in mapping[cur]:
                    indegree[nxt] -= 1
                    removed_edges += 1
                    if indegree[nxt] == 0:
                        q.append(nxt)
        return ans[::-1] if removed_edges == total_edges else []

File: Graphs/LC_210_Course_Schedule_II.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.18

class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        // topology sorting using BFS
        int n = numCourses;
        vector<int> indegrees(n, 0);
        int total_edges = 0, removed_edges = 0;
        unordered_map<int, vector<int>> graph;
        for (int i = 0; i < prerequisites.size(); i++){
            int cur = prerequisites[i][1], prev = prerequisites[i][0];
            indegrees[cur]++;
            graph[prev].push_back(cur);
            total_edges++;
        }
        queue<int> q;
        for (int i = 0; i < n; i++){
            if (indegrees[i] == 0){
                q.push(i);
            }
        }
        vector<int> ans;
        while(!q.empty()){
            int cur = q.front();
            q.pop();
            ans.push_back(cur);
            for(int nxt: graph[cur]){
                removed_edges++;
                indegrees[nxt]--;
                if (indegrees[nxt] == 0){
                    q.push(nxt);
                }
            }
        }
        if (removed_edges == total_edges){
            reverse(ans.begin(), ans.end());
            return ans;
        }
        return {};
    }
};